pallets/gear-eth-bridge/src/lib.rs

```diff
@@ -295,16 +341,113 @@ pub mod pallet {
             origin: OriginFor<T>,
             destination: H160,
             payload: Vec<u8>,
+            value: BalanceOf<T>,
         ) -> DispatchResultWithPostInfo {
             let source = ensure_signed(origin)?.cast();
 
+            Self::try_charge_fee(source, value, false)?;
             Self::queue_message(source, destination, payload)?;
 
             Ok(().into())
         }
     }
 
-    impl<T: Config> Pallet<T> {
+    impl<T: Config> Pallet<T>
+    where
+        T::AccountId: Origin,
+    {
+        pub(crate) fn try_charge_fee(
+            source: ActorId,
+            value: BalanceOf<T>,
+            from_builtin: bool,
+        ) -> Result<(), Error<T>> {
+            let account_id = T::AccountId::from_origin(source.into());
+
+            // Charge fee if queue threshold exceeded
+            if Queue::<T>::get().len() >= <T as Config>::QueueFeeThreshold::get() as _ {
+                let fee = <T as Config>::MessageFee::get();
+                // Return error if fee is greater than value
+                let refund = value.checked_sub(&fee).ok_or(Error::<T>::InsufficientFee)?;
+
+                AccountsFee::<T>::mutate(account_id.clone(), |details| {
+                    let details = details.get_or_insert_with(Default::default);
+                    details.fee = details.fee.saturating_add(fee);
+                    details.refund = details.refund.saturating_add(refund);
+                });
+            }
+            // Queue threshold not exceeded, no fee charged, we should make full value refund
```

Comment:
Sorry for this insufficient remark. However I left another one so should be ok ).

I would place that comment and the previous one within the corresponding block.

Reply:
These comments describe what happens inside the corresponding block. In this case, I prefer to place them just before the block rather than inside it.